File: data_structures/linked_lists/detect_a_cycle/detect_a_cycle.py
"""
:problem: https://www.hackerrank.com/challenges/ctci-linked-list-cycle/problem
"""


class Node(object):
    def __init__(self, data=None, next_node=None):
        self.data = data
        self.next = next_node


def has_cycle(head: Node):
    fast = head
    slow = head

    while fast and fast.next:
        fast = fast.next.next
        slow = slow.next
        if fast is slow:
            return True

    return False

File: data_structure/linked_list/detect_a_cycle/detect_a_cycle.cpp
/**
 * @problem https://www.hackerrank.com/challenges/ctci-linked-list-cycle/problem
 */

struct Node {
    int data;
    struct Node* next;
};

bool has_cycle(Node* head) {
    Node* fast = head;
    Node* slow = head;

    while (fast && fast->next) {
        fast = fast->next->next;
        slow = slow->next;
        if (fast == slow) {
            return true;
        }
    }
    return false;
}
